### ISSUE TITLE
Feature request - allow "zfs recv" on remote host executed with sudo to avoid connecting as root

### ISSUE THREAD

**Issue by homonto** [gbytedev/zfsbud#39]
Hi (or: "czesc Pawel"),

first of all: marvelous piece of work - I just was looking for something like this to allow replication between TrueNAS and non-TrueNAS systems - and it works
but now I have a proposal: "zfs recv" will only work if you execute it as root (at least on TrueNAS)
that however brings security issue: you need to ssh root@remote-system
rather than that I would propose:
ssh user-with-privilidges@remote-host 
and then:
sudo zfs recv

this of course requires changing the script i.e. to specify "--sudo" - use sudo on remote host

what do you think?

I am testing it now by modyfing your script but I don't want to destroy it - rather it would be better via PR or so (I have never done PR yet in my life so let me see what I can do)

Thank you
Dziekuje

**Comment by gbytedev** [comment]
> but now I have a proposal: "zfs recv" will only work if you execute it as root (at least on TrueNAS)

I'm not sure about TrueNAS but genereally you would add certain ZFS permissions to a non-root user; e.g. on the remote system you would execute `sudo zfs allow some-user create,receive,... some-pool`, see [zfs allow](https://docs.oracle.com/en/operating-systems/solaris/oracle-solaris/11.4/manage-zfs/delegating-zfs-permissions-zfs-allow.html). Does hat help your use case?

**Comment by homonto** [comment]
thank you, I just tested this and that and it is kind of convoluted
why?
if I do: "zfs allow -u user mount pool" then it is still an issue because user can only mount when he has enough file system privilidges to the mount point
while doing all zfs commands as root or sudo, all these problems are non existing
so in short:
1- manipulating "zfs allow...." can end up with very difficult permissions on zfs/filesystem
2- adding "sudo zfs..." to your script would solve it all

that is me thinking and I might be wrong as inexperienced zfs user ;-)

**Comment by homonto** [comment]
plus: messing on the filesystem level with TrueNAS is no-no ;-)

**Comment by gbytedev** [comment]
AFAIK mounting is a special case on Linux but all(?) the other permissions are ZFS specific and do not need sudo after setting them via 'zfs allow'. This has nothing to do with your filesystem permissions. 

Allowing a regular user to create snapshots, receive etc. is the usual way of using zfs. It is very powerful as your receiving end does not need e.g. 'send' as permission and so you can use 'ZFS allow' to secure the box by specifying only what you need.

Tl;dr don't let root handle zfs operations and avoid sudo where possible. Use 'ZFS allow' to restrict who can do what on each host. I don't know about TrueNas but pretty sure it's not much different.

**Comment by homonto** [comment]
I actually agree with you FULLY, but...
TrueNAS is an appliance and ANY modification is first of all not welcome, but more importantly: it is WANISHED with first update of TrueNAS (that includes any permission given to users)
so either I workaround this script or mess up with production TrueNAS (and many of them) - of course you know the path I will take
btw I have done the modifications, added "-S|--sudo" argument and changed respectively the "zfs" command sot "sudo zfs" but only on the remote commands and only when used with "--rsh|-e"

again: thank you for your work
